tests/test_generation_utils.py

```diff
@@ -1903,3 +1903,122 @@ def test_generate_encoder_outputs_attention_mask(self):
         output_sequences_with_mask = output_sequences_with_mask.cpu()
 
         self.assertListEqual(output_sequences_no_mask.tolist(), output_sequences_with_mask.tolist())
+
+    @staticmethod
+    def compute_transition_beam_scores_sum(outputs, vocab_size):
+        # reshape scores
+        scores = torch.stack(outputs.scores).reshape(len(outputs.scores), -1).transpose(0, 1)
+        cut_idx = outputs.sequences.shape[-1] - scores.shape[-1]
+        # compute real indices
+        indices = (
+            outputs.sequences[:, cut_idx:]
+            + torch.tensor(outputs.beam_indices, device=outputs.sequences.device) * vocab_size
+        )
+
+        # gather scores and run
+        return scores.gather(0, indices).sum(-1)
```

Comment:
This seems like a core algorithm so IMHO it could use some code clarification.
Also in that way you don't need to pass `vocab_size` externally.
And that function seems really useful (without the final sum).

Couldn't we instead make it part of `BeamDecoderOutput` or a raw function in `generation_utils` ?
Seems to me at least that the returned tensor is what I would expect roughly from `outputs.scores` (maybe `outputs.generated_scores()` to not break BC ?)

```suggestion
    def compute_transition_beam_scores_sum(outputs):

        # reshape scores
        # output.scores = [seq_len, batch_size, vocab_size]
        scores = torch.stack(outputs.scores)
        (seq_len, batch_size, vocab_size) = scores.shape
        scores = scores.reshape(seq_len, -1).transpose(0, 1)
        # scores = [vocab_size * batch_size, seq_len]

        # The output sequences might contain some decoder_initial_ids. Ignore those.
        cut_idx = outputs.sequences.shape[-1] - scores.shape[-1]

        # compute real indices
        beam_offsets = torch.tensor(outputs.beam_indices, device=outputs.sequences.device) * vocab_size
        indices = outputs.sequences[:, cut_idx:] + beam_offsets

        # gather scores and run
        return scores.gather(0, indices).sum(-1)
```

Reply:
This (without the sum) is much more likely to be what a user wants than `scores`. Imho, it should be included with the Beam Search Output rather than `scores` and `beam_indices`, both of which are beam search internal values, whereas the per-token output probabilities are not.

I would not include the sum in this function, as the user already gets the sequence score as a separate output. The reason why the user is giving `return_scores=True` is because they want per-token logprobs.

How would this function (and `sequences_scores`) deal with return sequences of unequal length? Is the output after EOS filled with pad and logprob `-inf`? Or could there be garbage output and arbitrary probabilities?